[WIP] Handle 100 (Continue) HTTP response gracefully regardless of whether it was requested/expected

johan12345 opened the pull request:
I created a failing test for issue #3628 (first commit) - or rather changed the previous `serverRespondsWithUnsolicited100Continue` test that asserts that the 100 response would just be returned to the user. I also tried to implement a solution, i.e. make OkHttp send the response body again if a response with code 100 was received (second commit).

When testing it with a real-world server in our app where the problem previously occurred, the issue also seems to be fixed.

**However**, the next HTTP request (in this case it's a GET) after the request where the server sent the 100 response gets a 403 Forbidden error. I suspect this has something to do with the HTTP connection being in an inconsistent state so that it cannot be reused. But I don't really understand why this should be the case here (the client has sent the request body and read the response, so the request should in principle be done, right?). Does someone have an idea about this?

/cc @yschimke

yschimke commented on okhttp/src/main/java/okhttp3/internal/http1/Http1Codec.java:
```diff
@@ -196,6 +196,9 @@ public void writeRequest(Headers headers, String requestLine) throws IOException
 
       if (expectContinue && statusLine.code == HTTP_CONTINUE) {
         return null;
+      } else if (statusLine.code == HTTP_CONTINUE) {
```
I don't understand the need for this branch since it duplicates the lines below

johan12345 replied:
It sets the state to a different one (STATE_OPEN_REQUEST_BODY instead of STATE_OPEN_RESPONSE_BODY). Otherwise, sending the response body again would not work.

yschimke replied:
It's early, sorry!

yschimke commented on okhttp/src/main/java/okhttp3/internal/http/CallServerInterceptor.java:
```diff
@@ -99,11 +90,31 @@ public CallServerInterceptor(boolean forWebSocket) {
       response = response.newBuilder()
-          .body(Util.EMPTY_RESPONSE)
-          .build();
+              .body(Util.EMPTY_RESPONSE)
```
nit: inconsistent with formatting 3 lines below

yschimke commented on okhttp/src/main/java/okhttp3/internal/http/CallServerInterceptor.java:
```diff
@@ -99,11 +90,31 @@ public CallServerInterceptor(boolean forWebSocket) {
 
     int code = response.code();
+    if (code == 100) {
+      // server sent a 100-continue even though we did not request one. Send the request body again.
```
This doesn't seem to match the RFC which states

> The client ought to continue sending the request and discard the 100 response.

Specifically, why are we resending the request body? It seems like we should ignore the 100 and just try to read a later HTTP response?

johan12345 replied:
Good point. I will try if it works without resending the body.

johan12345 commented:
@yschimke You're right, it was not necessary to re-send the request body, just to read the response again. I updated the PR accordingly.

In the case that the server just sends a 100 response, but not the actual response, the request now times out. I changed my commit to the CallTest so that this behavior is also tested.

johan12345 commented:
Is there anything else I need to do here or can this be merged? Also, do you plan to release a new OkHttp version soon or should we build a JAR ourselves to fix this problem in our app in the meantime?

swankjesse commented on okhttp/src/main/java/okhttp3/internal/http/CallServerInterceptor.java:
```diff
@@ -99,6 +99,25 @@ public CallServerInterceptor(boolean forWebSocket) {
         .responseHeadersEnd(realChain.call(), response);
 
     int code = response.code();
+    if (code == 100) {
```
In follow up we should probably check the entire range.

johan12345 replied:
So you mean allowing this for every 1xx status code instead of just 100? Yes, that should be possible.

swankjesse commented on okhttp/src/main/java/okhttp3/internal/http/CallServerInterceptor.java:
```diff
@@ -99,6 +99,25 @@ public CallServerInterceptor(boolean forWebSocket) {
         .responseHeadersEnd(realChain.call(), response);
```
I'm pondering the consequences of calling this 2x. It's going to crash listeners.

Can you change your fix to go before this listener call back? And we won't call back on the headers for the unsolicited 100; it's not worth the trouble.

johan12345 replied:
Okay, I changed this accordingly in 785e1d4

swankjesse commented:
Sorry for the delay. I've been busy.

swankjesse commented on okhttp/src/main/java/okhttp3/internal/http/CallServerInterceptor.java:
```diff
@@ -95,10 +95,26 @@ public CallServerInterceptor(boolean forWebSocket) {
+    if (code == 100) {
+      // server sent a 100-continue even though we did not request one.
+      // try again to read the actual response
+      realChain.eventListener().responseHeadersStart(realChain.call());
```
we still have unbalanced calls to responseHeadersStart(). Mind dropping this call too?

johan12345 replied:
Oh, right. Fixed in 440531d.

yschimke commented:
@johan12345 did you fill in this contributor agreement?

https://github.com/square/okhttp/wiki/Contributing

johan12345 commented:
@yschimke Yes, I think I did this already shortly after submitting the PR.
